### PR TITLE
Move schedule cleanup from program to object

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Move schedule cleanup from program to object and add function to notify when ref-count reaches zero.

This is to resolve a race condition that is breaking kernel CI/CD.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/execution_context/ebpf_program.c [hunk 1]
```diff
@@ -234,15 +212,17 @@ _ebpf_program_epoch_free(void* context)
...
-
-    ebpf_free(program->cleanup_work_item);
     ebpf_free(program);
 }
 
+static const ebpf_program_type_t*
+_ebpf_program_get_program_type(_In_ const ebpf_object_t* object)
+{
+    return ebpf_program_type((const ebpf_program_t*)object);
+}
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:224]: Question: why was this moved down from line 194?

File: libs/platform/ebpf_object.c [hunk 2]
```diff
@@ -133,6 +133,19 @@ ebpf_object_tracking_terminate()
     }
 }
 
+/**
+ * @brief Free invoked when the current epoch ends. When the object ref-count reaches zero.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:137]: "When the object ref-count reaches zero" is not a complete sentence.  Not sure what was intended.

File: tests/sample/sample.vcxproj [hunk 3]
```diff
@@ -141,16 +141,6 @@
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">$(OutputPath)%(Filename).o</Outputs>
       <Outputs Condition="'$(Configuration)|$(Platform)'=='Release|x64'">$(OutputPath)%(Filename).o</Outputs>
     </CustomBuild>
-    <CustomBuild Include="encap_reflect_packet.c">
```
Comment (dthaler) [inline tests/sample/sample.vcxproj:144]: Why is this one removed?  Seems unrelated to moving schedule cleanup from program to object.

Comment (Alan-Jowett) [reply tests/sample/sample.vcxproj:144]: The project fails to load.

File: libs/platform/ebpf_object.h [hunk 4]
```diff
@@ -69,10 +74,11 @@ extern "C"
      */
     ebpf_result_t
     ebpf_object_initialize(
-        ebpf_object_t* object,
+        _In_ ebpf_object_t* object,
         ebpf_object_type_t object_type,
-        ebpf_free_object_t free_function,
-        ebpf_object_get_program_type_t get_program_type_function);
+        _In_opt_ ebpf_zero_ref_count_object_t zero_ref_count_function,
```
Comment (dthaler) [inline libs/platform/ebpf_object.h:79]: avoid abbreviations ("ref") per style guidance

File: libs/execution_context/ebpf_maps.c [hunk 5]
```diff
@@ -959,14 +959,26 @@ ebpf_map_function_table_t ebpf_map_function_tables[] = {
...
+ *
+ * @param[in] object Pointer to ebpf_object_t whose ref-count reached zero.
+ */
 static void
-_ebpf_map_delete(_In_ ebpf_object_t* object)
+_ebpf_map_zero_ref_count(_In_ ebpf_object_t* object)
 {
     ebpf_map_t* map = (ebpf_map_t*)object;
 
     if (map->inner_map_template != NULL) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/execution_context/ebpf_maps.c:972]: We should change this to drop all the internal reference counts immediately, not just the one held by the template.  (The others are currently released inside the delete map api for each relevant map type, such as `_delete_object_array_map()`.)

File: libs/execution_context/ebpf_maps.c [hunk 6]
```diff
@@ -959,14 +959,26 @@ ebpf_map_function_table_t ebpf_map_function_tables[] = {
...
 };
 
+/**
+ * @brief Invoked by ebpf_object_t reference tracking when the refcount reaches zero.
+ *
+ * @param[in] object Pointer to ebpf_object_t whose ref-count reached zero.
+ */
 static void
-_ebpf_map_delete(_In_ ebpf_object_t* object)
+_ebpf_map_zero_ref_count(_In_ ebpf_object_t* object)
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/execution_context/ebpf_maps.c:968]: How about we make a distinction between "delete" vs "free" and use those terms.  "delete" means to delete from all global tracking lists, etc, and happens when refcount hits 0.  "free" means to actually free the memory and happens when flushing an epoch and garbage collection is done.
